### PR TITLE
Upgrade libsnark to master

### PR DESCRIPTION
**Pull request by Schaeff**
Changes the build process to use submodules. Replaces #67 to preserve commit history.

Todo:

- [x] fix `release` build
- [x] fix non-docker development environment
- [x] add specific commit for libsnark

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -2,33 +2,39 @@ FROM ubuntu:14.04
...
 
-ENV LD_LIBRARY_PATH $LD_LIBRARY_PATH:/usr/local/lib
+WORKDIR /root/
 
-COPY . /root/ZoKrates
+COPY . ZoKrates
 
 RUN cd ZoKrates \
-  && cargo build --release
+    && cargo build
```
Comment (xwvvvvwx) [inline Dockerfile]: why remove the `--release`?

File: build.rs [hunk 2]
```diff
@@ -1,23 +1,42 @@
...
+            libsnark.join("lib").display()
+        );
+
         println!("cargo:rustc-link-lib=gmp");
         println!("cargo:rustc-link-lib=gmpxx");
-        println!("cargo:rustc-link-lib=supercop");
-        println!("cargo:rustc-link-lib=snark");
-        println!("cargo:rustc-link-search=/usr/local/lib");
+        println!("cargo:rustc-link-lib=static=snarkd");
+        println!("cargo:rustc-link-lib=static=ffd");
```
Comment (xwvvvvwx) [inline build.rs]: I think this will not work in release config?

File: Dockerfile [hunk 3]
```diff
@@ -2,33 +2,39 @@ FROM ubuntu:14.04
...
-    apt-get install -y \
-    wget unzip curl \
-    build-essential git libgmp3-dev libprocps3-dev libgtest-dev python-markdown libboost-all-dev libssl-dev
-
-RUN wget https://github.com/scipr-lab/libsnark/archive/$libsnarkcommit.zip \
-  && mv $libsnarkcommit.zip libsnark.zip \
-  && unzip libsnark.zip \
-  && cd libsnark-$libsnarkcommit \
-  && ./prepare-depends.sh
+ENV LD_LIBRARY_PATH $LD_LIBRARY_PATH:/usr/local/lib
```
Comment (xwvvvvwx) [inline Dockerfile]: Is this still required?

File: Dockerfile (hunk 1 above)
Comment (Schaeff) [reply Dockerfile]: good catch

File: Dockerfile (hunk 3 above)
Comment (Schaeff) [reply Dockerfile]: no, removing

File: build.rs [hunk 4]
```diff
@@ -1,23 +1,51 @@
...
+            .build();
+
         gcc::Build::new()
             .cpp(true)
-            .debug(true)
-            .include("/usr/local/include")
-            .include("/usr/local/include/libsnark")
+            .debug(cfg!(debug_assertions))
             .flag("-std=c++11")
+            .include(libsnark_source_path)
```
Comment (Schaeff) [inline build.rs]: @Kyroy should be fine now

**Comment by Schaeff** [comment]
Hey @Kyroy thanks for the updates! I think we can merge?

**Comment by Kyroy** [comment]
Before merging, consider the open PR #64

**Comment by Schaeff** [comment]
Now that our next release is in it's own `rc` branch, isn't it fine to continue pushing to develop for the next release?

**Pull request merged**